File: JavaScript/quickSort.js

function quickSort(a) {
  qSortRecursive(a,0,a.length-1);
}

function qSortRecursive(a,start,stop){
  if(start < stop){
    var boundary = partition(a,start,stop);
    qSortRecursive(a,start,boundary-1);
    qSortRecursive(a,boundary+1,stop)
  }
}

function partition(a,start,stop){
  var pivot = a[start];
  var left = start + 1;
  var right = stop;
  var done = false;
  var count = 0;
  while(!done)
  {
    count++;
    while(left <= right && a[left] <= pivot)
      left++;

    while(right >= left && a[right] >= pivot)
      right--;

    if(left >= right){
      done = true;
    }else{
      [a[left],a[right]] = [a[right],a[left]];
    }

    if(count > 15)
    {
      break;
    }
  }

  [a[start],a[right]] = [a[right],a[start]];
  return right;
}


mya = [23,45,1,2,-34,236,-11,4,25,-72,-1000,67,2,-34,45,78];
quickSort(mya);
console.log(mya);

File: Ruby/quickSort.rb
def quickSort(a)
  qSortR(a,0,a.length-1)
end

def qSortR(d,start,stop)
  if start < stop then
    boundary = partition(d,start,stop)
    qSortR(d,start,boundary-1)
    qSortR(d,boundary+1,stop)
  end
end

def partition(d,start,stop)
  pivot = d[start]
  left = start + 1
  right = stop
  done = false

  while not done do
    while left <= right and d[left] <= pivot do
      left += 1
    end

    while right >= left and d[right] >= pivot do
      right -= 1
    end

    if left > right then
      done = true
    else
      temp = d[left]
      d[left] = d[right]
      d[right] = temp
    end
  end


  d[start] = d[right]
  d[right] = pivot

  return right
end

arr = [6,2,87,-1,-3,0,1,1,-3,-43,23,67,56]
puts
quickSort(arr)
print(arr)
